### ISSUE TITLE
Settings to control resources usage: pushed downloads - cryptocurrency mining (?)

### ISSUE THREAD

**Issue by JazzTp** [webtorrent/webtorrent-desktop#1453]
## Is webtorrent-desktop, directly or through some additional code pushed-downloaded to it, mining some crypto currency? Did I miss something about it?

I'm taking the time to bother you because I think that webtorrent-desktop looks _gorgeous_ and very promising, although I'd have **a few suggestions to improve** its interface and usability, nothing complicated but really **indispensable IMO** in case one wants to add plenty of torrents to it. **However, this issue is not about that.**

**What version of WebTorrent Desktop?**
webtorrent-desktop_0.20.0-1_amd64

**What operating system and version?**
Linux Ubuntu 16.04

**What did you do?**
I added 35-40 torrents in order to test and contribute peer2peer bandwidth, the Blender Foundation ones and the ones of an independent info channel that I follow.

**What did you expect to happen?**
I expected that the webtorrent-desktop instance would just keep those files available on the network and serve them when requested, contributing bandwidth+storage.
https://webtorrent.io/desktop/

> Features
> + **_Lightweight_**, fast torrent app
> + Beautiful user experience
> + Free, non-commercial, ad-free, and open source
> + ...
> 

**What actually happened?**
+ Sometimes, _the whole Gnome session_ became unresponsive, I couldn't stop/kill webtorrent-desktop, it was even hard to type a terminal command to kill the whole Gnome session (as well as anything else running on top of it at the moment), or I had to switch off the machine and restart it. In other occasions, like today, all my GUI just started stuttering, it was well noticeable when dragging a window or when opening BitChute and playing a video to check it out and see if it's new material yet to be downloaded.
+ Computer heat raises, CPU and GPU fans go faster.
+ The `top` command shows very heavy load on the CPU.

**Again, the features list on the website says "lightweight". If mining is to support some project, it can be OK to many users, but I think this side of the client should be made clearly visible to users and it should be explained through the app GUI what the purpose is and where the benefits go.**

**_Of course_, developers (or the ones in control of the project) have the right to decide whether it's in the users' hands to control or a "take it or leave it" feature.**

I didn't see any settings about mining crypto through it, like
+ % CPU to use for mining (0 disables)
+ % GPU to use for mining (0 disables)

If it's all or nothing at all, a user can be obliged to opt for another client or totally abandon any will to contribute bandwidth and storage to BitChute, for various **obvious** reasons, for instance:
+ The machine owner might _have to_ keep something else going at the same time
+ Heat during summer
+ Electricity bill in certain countries/periods

The whole client could be launched with a low priority level probably, at least on Linux, but that would leave the torrent side of it also penalized.

**I _strongly_ suggest that settings regarding the use of resources are made easily accessible through the GUI.**

**I _strongly_ believe in distributed peer2peer hosting against the arbitrary censorship and arrogant behavior of platforms like YouTube, which certainly do not represent any champions of respect of democracy or human rights.**

**I think it is of _vital_ importance, for webtorrent-desktop to thrive, and for the success of webtorrent and platforms which adopted it like BitChute, that potential users are not facing anything which might not be crystal clear and which might arise - rightfully or by misunderstanding - suspects of any kind.**

---

## While we are talking about controlling resources usage: pushed downloads

A download was "pushed" to webtorrent-desktop soon after starting it up, just after adding those 35-40 entries.

Hadn't I noticed the ongoing download, I might have never noticed the presence of that material, because it appeared on the most ancient side of the list, it was electronic music, mentioned in issue #905 (I had no problems deleting it on Linux).

**I didn't see any options about allowing the download of "pushed" material**, like
+ always ask (default)
+ always accept
+ never accept
+ automatically accept when matches expression
+ automatically reject when matches expression

**Comment by DiegoRBaquero** [comment]
1. We don't do cryptocurrency mining.

2. "Pushed" torrents are loaded by default on first execution and start paused:

![image](https://user-images.githubusercontent.com/2006839/44165847-70eabe80-a08f-11e8-8a62-f7af0b4ff485.png)

3. That's the only "pushed" material to WD ever, all with open licenses. Easily removed imo.

Hope this answers your questions

**Comment by JazzTp** [comment]
@DiegoRBaquero 

1. I wonder what's using so heavily the CPU (according to `top`) and apparently the GPU too, and I wonder if any library or third-party code you might be using in webtorrent-desktop could be responsible for it.

2. OK on this one, just... at startup, on some _previously_ downloaded torrents for which it was _already_ showing "Seeding", it shows sometimes "Downloading" and later "Download Complete", I must have seen that + I hadn't "registered" the presence of that torrent before. It should probably display "Checking" instead of "Downloading". 

I see that you have already closed the issue, I guess you're very busy, so I won't bother you again now, I might try out some _future_ release.

**Comment by DiegoRBaquero** [comment]
@JazzTp if you are seeding lots of torrent, WebTorrent does consume a lot of CPU, keep in mind that WebTorrent, both lib and Desktop app are in BETA and lack many many optimizations, for example hashing of pieces is done in JS, which is very inefficient.

Try removing all torrents and see if CPU usage remains high. It should go to 1% or something really small, like an idle chrome.

2. This is a bug we are aware off, because we re check all the pieces on startup, but it's marked as downloading. Actually, there's a proposal to no recheck them and trust what is on disk as is.

We are all busy but I'm happy to answer any questions you have :)

**Comment by Borewit** [comment]
@JazzTp: be factual in the future. Keep your nonsense regarding crypto-mining for yourself until you are able to present solid evidence we are part of the conspiracy.

**Comment by JazzTp** [comment]
@DiegoRBaquero Counted: 37 torrents, and that CPU load is also measurable when actually serving none of them. I've thrown about 120 torrents into qbittorrent to test, and with 13 actively being served it uses about 60 times less CPU and puts no evident load on the GPU. I'm aware that qbittorrent is not in beta and that I'm _possibly_ comparing C++ vs. Java code...
The hash thing: but you do not keep recalculating the same hash codes again and again, right?

BTW while we are "talking" let me mention a suggestion without opening an issue or a feature request:


When adding a magnetic URL from BitChute, you end up getting the BitChute hash code on the interface as the only text associated to that torrent. You really want to be able to manually input a title (if not also a file name) and chose a subfolder where to save.


@Borewit  **LOL I _am_ aware that there were various cases of embedded crypto-mining code, µTorrent for instance, but this one is so evident that it is _impossible_ to suspect a conspiration, it would be the worst ever.**

Honestly, one possibility I was considering is that some young developers living full immersion in the code where trying new stuff that their community might be aware of, but without adding the related settings in the GUI and without any mentions in the documentation or on the app website, so somebody coming from outside like me and not following the community developments would not know that the thing is being discussed (I've seen something about crypto mining related to BitChute in the forum, didn't dig into it honestly... _yes_ I'm aware that BitChute _uses_ webtorrent but is _not_ the same as the developers of webtorrent-desktop, it might have been related, or not).

**I came here asking _if_ that was the case** and if that was the case I was meaning to _urge_ you to mention those features and put something well visible in the GUI ASAP and possibly make the related settings accessible (GUI or text config file to edit or whatever), in order to stay crystal clear _before_ anybody might come out with "the webtorrent thing is taking advantage of users' resources to mine without their knowledge". **Because as I said I believe webtorrent is a great thing and webtorrent-desktop looks very promising. Actually, you might have read very quickly here and there my post because it was too long (I would actually like _not_ to feel that I have to be writing all this stuff now, either). I _had_ also mentioned a possible misunderstanding.**

Well if you don't read the whole thing I can't complain because it was long, but in turn, don't complain. Or **in doubt, think positive.**

**As for "factual":** honestly, the symptoms I described are factual enough IMO, and I answered all the questions your input form poses including the OS type and version. I was going to include a snapshot of `top` with webtorrent-desktop and qbittorrent and other processes running, for you to compare, then I thought "let them find out, the number is so high, don't be even more pedant, everybody can launch `top` to see the CPU load after adding about that number of torrents instead of only testing with the Blender Foundation ones, and then be able to continue testing with any code tweak they might think of".

Not factual were my other hypothesis, so I omitted all of them, I had thought about some possible resouces-sucking continuous polling flow instead of maybe an event driven flow (provided that would be possible, I'm frankly speaking no expert at all in network layer programming or protocol level programming), but whatever fantasy I could come out with, it was not justifying the _apparent_ GPU load. I get that _apparent_ load - it actually stays much more responsive - when rendering with Cycles in Blender, GPU only (with even less loss of response with the more recent GPU+CPU Cycle engine or with Eevee).

**_Sorry_ that you have taken it the offensive way**, if you did. After posting I suspected for a short while that it might happen, but job-related people arrived and I had no chance to come back immediately for substantial reedit. As mentioned in certain rules of other technical exchange websites, it's difficult to decipher tone in text. **In doubt, I think positive** (so I laughed at your "conspiracy" thing but maybe you were furious while writing it and meant no joke at all, LOL).

Keep on the great work. I'm happy that you didn't do the crypto thing without mentioning it. You might want to put a note on your website and/or download page emphasizing "This is still a beta version and possibly not a really lightweight one, _yet_. Code optimization is still being done." or "... Much code optimization is still due." or something like that.
